[GPU Snapshot] Add Clear History Flag

sraikund16 opened the pull request:
Summary:
Oftentimes, users complain that a bunch of extra events are prepended to their desired GPU snapshot. This is because they usually attach an OOM logger without knowing and when they go to collect the actual snapshot, it adds all the OOM logger contents. Since OOM and regular snapshot use the same backend, we currently don't have the infra in place to split these snapshots.

As a solution we add a flag to the snapshot frontend to clear out the history when starting the auto-trace record memory history.

A more thorough solution would be to have a user pass in a handle and to have snapshots per handle to seperate the events. However, this would likely be complicated and more work than it is worth as we would have to change the callbacks in the caching allocator and pass these objects between python and cpp.

Test Plan:
See diff below

Differential Revision: D71159720




cc @ptrblck @msaroufim @eqy

sraikund16 commented:
@eqy @syed-ahmed Can I get a review?

eqy reviewed (approved):
no objections, but seems like you need to update https://github.com/pytorch/pytorch/blob/3b48c7214179057fa6457e5049b20999604f8acc/torch/_C/__init__.pyi.in#L1919 ?

sraikund16 commented:
> no objections, but seems like you need to update
> 
> https://github.com/pytorch/pytorch/blob/3b48c7214179057fa6457e5049b20999604f8acc/torch/_C/__init__.pyi.in#L1919
> 
> ?

Oops forgot about linter. Thank you!

sraikund16 committed:
[GPU Snapshot] Add Clear History Flag (#149352)

Summary:

Oftentimes, users complain that a bunch of extra events are prepended to their desired GPU snapshot. This is because they usually attach an OOM logger without knowing and when they go to collect the actual snapshot, it adds all the OOM logger contents. Since OOM and regular snapshot use the same backend, we currently don't have the infra in place to split these snapshots.

As a solution we add a flag to the snapshot frontend to clear out the history when starting the auto-trace record memory history.

A more thorough solution would be to have a user pass in a handle and to have snapshots per handle to seperate the events. However, this would likely be complicated and more work than it is worth as we would have to change the callbacks in the caching allocator and pass these objects between python and cpp.

Test Plan:
Without clear history:
https://www.internalfb.com/pytorch_memory_visualizer/ai_efficiency/tree/gpu_snapshot/devvm2185.cco0.facebook.com/rank-0_itrn-13.Mar_17_14_42_46.2361636.snapshot.pickle

With clear history:
https://www.internalfb.com/pytorch_memory_visualizer/ai_efficiency/tree/gpu_snapshot/devvm2185.cco0.facebook.com/rank-0_itrn-13.Mar_17_15_06_32.2746734.snapshot.pickle

Reviewed By: aaronenyeshi

Differential Revision: D71159720
